File: 20-ValidParentheses/20-ValidParentheses.py
# Last updated: 3/25/2026, 9:07:58 AM
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        
        for c in s:
            # Push opening brackets
            if c in "({[":
                stack.append(c)
            else:
                # No opening bracket to match
                # Checks whether the stack is empty before trying to remove (pop) an element from it.
                if len(stack) == 0:
                    return False
                
                top = stack.pop()
                
                # Check matching pairs
                if (c == ')' and top != '(') or \
                   (c == '}' and top != '{') or \
                   (c == ']' and top != '['):
                    return False
        
        # Valid only if no unmatched brackets remain
        return len(stack) == 0

File: 20-ValidParentheses/20-ValidParentheses.cpp
// Last updated: 3/25/2026, 9:08:01 AM
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;

        for(char c : s) {
            if(c == '(') st.push(')');
            else if (c == '{') st.push('}');
            else if (c == '[') st.push(']');
            else {
                if (st.empty() || st.top() != c) return false;
                st.pop();
            }
        }
        if(!st.empty()) return false;
        return true;
        
        // for (char c : s) {
        //     // Push opening brackets onto stack
        //     if (c == '(' || c == '{' || c == '[') {
        //         st.push(c);
        //     }
        //     // Check for closing brackets
        //     else {
        //         if (st.empty()) return false; // no opening bracket to match
                
        //         char top = st.top();
        //         st.pop();
                
        //         if ((c == ')' && top != '(') ||
        //             (c == '}' && top != '{') ||
        //             (c == ']' && top != '[')) {
        //             return false;
        //         }
        //     }
        // }
        
        // return st.empty(); // valid only if no unmatched brackets remain
    }
};